### PR TITLE
Blocking queue for reader

### PR DESCRIPTION
**Pull request by JiayiFeng**
This PR replaces the invocations of `framework::Channel` in readers with new implementation of `reader::BlockingQueue`.

#### Why not keep using `framework::Channel`? 
`framework::Channel` is a great idea. It provides rich functionality and the usage of only a small part of it has made the developing of readers quite easy. However, the implementation of `framework::Channel` is not very stable so far. It crashes in GCC 4.8.2 environment and suffers occasional deadlock. So we implemented the `reader::BlockingQueue`, which can be regarded as an extremely simplified `framework::Channel`. It has similar interfaces with `framework::Channel` while only provides features that readers really need. Its conciseness makes itself easy to maintain, and also it runs a little faster than `frameowrk::Channel` (892s vs 910s, 1000 batches in the transformer job).

#### Why not merge `framework::BlockingQueue` and `reader::BlockingQueue`?
These two blocking queues have different customized features. The `framework::BlockingQueue` is mainly used in the `ParallelExecutor`, so it supports `extend` operation and timeout mechanism. The `framework::BlockingQueue` is only used in `Readers`. As a replacement of `framework::Channel`, it supports capacity limitation and closing mechanism. 
It's hard to implement all these features in a single blocking queue.

### CODE REVIEW

**Comment by tpatejko** [comment]
@JiayiFeng just a quick question. I noticed that there is already an implementation of blocking queue in fluid: `fluid/framework/blocking_queue.h`.

What is the difference between the blocking queue in this PR and the blocking queue that is already used in PaddlePaddle?

**Comment by JiayiFeng** [comment]
Hi @tpatejko, I have just updated the PR description. And maybe your questions can be answered by it.

**Comment by wangkuiyi** [comment]
> This PR replaces framework::Channel in C++ readers with new implementation of reader::BlockingQueue.

It seems that this PR DOESN'T replace framework::Channel. Instead, it replaces the **invocations** of framework::Channel in reader/*.

File: paddle/fluid/operators/reader/blocking_queue.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    std::unique_lock<std::mutex> lock(mutex_);
+    receive_cv_.wait(lock, [&] { return !queue_.empty() || closed_; });
+    if (!queue_.empty()) {
+      PADDLE_ENFORCE_NOT_NULL(elem);
+      *elem = queue_.front();
+      queue_.pop_front();
+      send_cv_.notify_one();
+      return true;
+    } else {
+      PADDLE_ENFORCE(closed_);
```
Comment (wangkuiyi) [inline paddle/fluid/operators/reader/blocking_queue.h]: 这里的假设是不是 —— “如果队列空，就必须是已经closed的”？

如果是，那么这个假设不合理吧。队列刚创立的时候是空的；此时，如果在writer写入数据之前，reader调用了Receive，难道就crash了？

File: paddle/fluid/operators/reader/blocking_queue.h [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+#include <deque>
+
+#include "paddle/fluid/platform/enforce.h"
+
+namespace paddle {
+namespace operators {
+namespace reader {
+
+template <typename T>
+class BlockingQueue {
```
[the remark this answers is not in the document]
Comment (JiayiFeng) [reply paddle/fluid/operators/reader/blocking_queue.h]: Done. Thanks!

File: paddle/fluid/operators/reader/create_double_buffer_reader_op.cc [hunk 3]
```diff
@@ -58,7 +58,7 @@ class DoubleBufferReader : public framework::DecoratedReader {
   bool HasNext() const;
 
   void StartPrefetcher() {
-    channel_ = framework::MakeChannel<size_t>(kChannelSize);
+    channel_ = new reader::BlockingQueue<size_t>(kChannelSize);
```
[the remark this answers is not in the document]
Comment (JiayiFeng) [reply paddle/fluid/operators/reader/create_double_buffer_reader_op.cc]: I think `channel` is a better name. It is implemented with blocking queue, while it is essentially a channel.

File: paddle/fluid/operators/reader/blocking_queue.h [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    send_cv_.notify_all();
+    receive_cv_.notify_all();
+  }
+
+  bool IsClosed() {
+    std::lock_guard<std::mutex> lock(mutex_);
+    return closed_;
+  }
+
+  bool CanSend() {
```
[the remark this answers is not in the document]
Comment (JiayiFeng) [reply paddle/fluid/operators/reader/blocking_queue.h]: It's correct that `CanSend` and `CanReceive` are not thread-safe. They are offered for two reasons:

1. `framework::Channel` has these two interfaces.  As a replacement of `framework::Channel`, `reader::BlockingQueue` would better to have similar interfaces. It is possible that we reuse the `framework::Channel` in readers in the future. Similar interfaces can reduce the migration workload.

2. In current implementations of all c++ readers, `CanSend` and `CanReceive` are invoked single-threaded. So no bug is caused.

File: paddle/fluid/operators/reader/blocking_queue.h [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+      receive_cv_.notify_one();
+      return true;
+    }
+  }
+
+  bool Send(T&& elem) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    send_cv_.wait(lock, [&] { return queue_.size() < capacity_ || closed_; });
+    if (closed_) {
+      return false;
```
[the remark this answers is not in the document]
Comment (JiayiFeng) [reply paddle/fluid/operators/reader/blocking_queue.h]: Great idea!

File: paddle/fluid/operators/reader/blocking_queue.h [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    } else {
```
[the remark this answers is not in the document]
Comment (JiayiFeng) [reply paddle/fluid/operators/reader/blocking_queue.h]: Done

File: paddle/fluid/operators/reader/reader_blocking_queue_test.cc [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+}
+
+TEST(BlockingQueue, CanSendTest) {
+  size_t cap = 1;
+  BlockingQueue<int> q(cap);
+  q.Send(1);
+  EXPECT_FALSE(q.CanSend());
+}
+
+void FirstInFirstOut(size_t queue_cap, size_t elem_num, size_t send_time_gap,
```
[the remark this answers is not in the document]
Comment (JiayiFeng) [reply paddle/fluid/operators/reader/reader_blocking_queue_test.cc]: In unit tests, sender threads start before receiver threads. If we don't set the `time_gap`, all sender threads may have finished before receiver threads starting. Bugs that only appear when senders and receivers run concurrently can not be found.

File: paddle/fluid/operators/reader/blocking_queue.h (hunk 1 above)
Comment (JiayiFeng) [reply paddle/fluid/operators/reader/blocking_queue.h]: 上面有一个条件变量的wait操作。所以当代码运行到这个`if...else...`的时候，意味着“队列非空或者队列已被close”。然后又进入了`else`分支，说明此时“队列为空”，那么显然，队列已经被close。

您说的一上来直接就receive的情况，这时候队列为空，并且没有close，那么会在前面的条件变量wait处就被阻塞掉。所以并不会出现您说的问题。

Comment (wangkuiyi) [reply paddle/fluid/operators/reader/blocking_queue.h]: 明白了！

**Comment by tpatejko** [comment]
@JiayiFeng @wangkuiyi Thanks for clarifying the idea!

**Pull request merged**